qiskit/circuit/library/generalized_gates/linear_function.py

```diff
@@ -0,0 +1,118 @@
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Linear Function."""
+
+from typing import Union, List, Optional
+import numpy as np
+from qiskit.circuit import QuantumCircuit, Gate
+from qiskit.circuit.exceptions import CircuitError
+
+
+class LinearFunction(Gate):
+    """An n_qubit linear circuit."""
+
+    def __init__(
+        self,
+        linear: Union[List[List[int]], np.ndarray, QuantumCircuit],
+        validate_input: Optional[bool] = False,
+    ) -> None:
+        """Create a new linear function.
+
+        Internally, represents a linear function acting on n qubits as a n x n matrix of 0s and 1s.
+
+        Args:
+            linear (list[list] or ndarray or QuantumCircuit):
+                either an n x n matrix, describing the linear function,
+                or a quantum circuit composed of linear gates only
+                (currently supported gates are CX and SWAP).
+
+            validate_input: if True, performs more expensive input validation checks,
+                such as checking that a given n x n matrix is invertible.
+
+        Raises:
+            CircuitError: if the input is invalid:
+                either a matrix is non {square, binary, invertible},
+                or a quantum circuit contains non-linear gates.
+        """
+        if not isinstance(linear, (list, np.ndarray, QuantumCircuit)):
+            raise CircuitError(
+                "A linear function must be represented either by a list, "
+                "a numpy array, or a quantum circuit with linear gates."
+            )
+
+        if isinstance(linear, QuantumCircuit):
+            # The following function will raise a CircuitError if there are nonlinear gates.
+            self._mat = _linear_quantum_circuit_to_mat(linear)
+
+        else:
+            # Check that the matrix is square
+            if not all(len(row) == len(linear) for row in linear):
+                raise CircuitError("A linear function must be represented by a square matrix.")
+
+            # Check that all entries are either 0 or 1
+            for row in linear:
+                for entry in row:
+                    if entry not in (0, 1):
+                        raise CircuitError("A linear function must only contain 0s or 1s.")
+
+            # Optionally, check that the matrix is invertible
+            if validate_input:
+                det = np.linalg.det(linear) % 2
+                if not np.allclose(det, 1):
+                    raise CircuitError(
+                        "A linear function must be represented by an invertible matrix."
+                    )
```

Comment:
In order to be a `Gate`, the matrix needs to be unitary, not just invertible.  It'd probably be better to use the `quantum_info` tests for that if necessary, but I think the test you need is actually just whether it's a permutation matrix, which is far simpler.

Reply:
@jakelishman - please note that `LinearFunction` internal structure contains an `n x n` boolean matrix representing an `n`-qubit linear reversible gate (e.g. containing only `CX` and `SWAP` gates),  and it is **not** a `2^n x 2^n` unitary matrix.
If this matrix is invertible, then one can always synthesize it into a `QuantumCircuit` (containing only `CX` and `SWAP` gates) using a Gaussian elimination process, which is described in the paper of Patel-Markov-Hayes: 
https://arxiv.org/abs/quant-ph/0302002 
See the synthesis code in: https://github.com/Qiskit/qiskit-terra/blob/9cc9225f249d3a895d9f3944175ea7b681d248c0/qiskit/transpiler/synthesis/graysynth.py#L183

Reply:
Thanks! I have added an example to the linear function description that hopefully shows the difference between a linear function and say a permutation.